src/line_launcher/mod.rs

```diff
@@ -0,0 +1,102 @@
+use midir::MidiOutputConnection;
+use std::sync::mpsc::Receiver;
+use wmidi::{Channel, MidiMessage, Note, Velocity};
+
+use crate::BeatNumber;
+
+enum State {
+    NotPlaying,
+    Playing { next_note_index: usize },
+    WaitingToFireFinalNoteOff,
+}
+
+const CHANNEL: Channel = Channel::Ch1;
+const VELOCITY: u8 = 100;
+
+pub struct LineLauncher {
+    beat_message_receiver: Receiver<BeatNumber>,
+    state: State,
+    notes: Vec<u8>,
+    output: MidiOutputConnection,
+}
+
+impl LineLauncher {
+    pub fn new(beat_message_receiver: Receiver<BeatNumber>, output: MidiOutputConnection) -> Self {
+        Self {
+            beat_message_receiver,
+            state: State::NotPlaying,
+            notes: vec![60, 53, 55, 58, 60, 61, 63, 65, 64],
+            output,
+        }
+    }
+
+    pub fn listen(&mut self) -> () {
+        loop {
+            let beat_message = self.beat_message_receiver.recv().unwrap();
+            match self.state {
+                State::NotPlaying
+                    if beat_message.quarter_note == 1 && beat_message.sixteenth_note == 1 =>
+                {
+                    self.state = State::Playing { next_note_index: 0 };
+                    self.play_note();
+                }
+                State::Playing { next_note_index } => {
+                    if next_note_index > 0 {
+                        self.fire_note_off(next_note_index - 1);
+                    }
+                    self.play_note();
+                }
+                State::WaitingToFireFinalNoteOff if beat_message.sixteenth_note == 1 => {
+                    self.fire_note_off(self.notes.len() - 1);
+                    self.state = State::NotPlaying
+                }
+                _ => {}
+            }
+        }
+    }
+
+    fn play_note(&mut self) -> () {
+        match self.state {
+            State::Playing { next_note_index } => {
+                self.fire_note_on(next_note_index);
+
+                self.state = if next_note_index == self.notes.len() - 1 {
+                    State::WaitingToFireFinalNoteOff
+                } else {
+                    State::Playing {
+                        next_note_index: next_note_index + 1,
+                    }
+                };
+            }
+            _ => (),
```

Comment:
This is clunky, there's an invariant that this method should only get called when we're in `State::Playing` but not sure what's the nicest way to express that?

Reply:
In general, the way I would do that would be to have the `Playing` state hold some kind of struct which exposed this function, so that it's not even possible to call it when you're in another state. I'm not sure if that could work here.